[[ WindowsPlayer ]] Update the player object on Windows to use DirectShow

livecodeian opened the pull request:
Modified MCPlatformPlayer to return the native view, using MCNativeLayer to manage geometry / visibility changes.
Implemented new MCPlatformPlayer based on DirectShow - switched Windows engine to use MCPlayer based on MCPlatformPlayer.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/livecode/livecode/4022)

<!-- Reviewable:end -->

livecodeian committed:
[[ WindowsPlayer ]] [[ WindowsPlayer ]] Add methods to get video rect from player rect and vice versa.

livecodeian committed:
[[ WindowsPlayer ]] Add getdevicetransform() method to MCContext

livecodeian committed:
[[ WindowsPlayer ]] Add image size parameters to MCPlatformLockPlayerBitmap and bool return value to indicate success.
[[ WindowsPlayer ]] Modify MCPlayer::draw() to pass required image size when locking bitmap for offscreen rendering.

livecodeian committed:
[[ WindowsPlayer ]] Move management to video view to MCNativeLayer class

livecodeian committed:
[[ WindowsPlayer ]] Remove call to MCPlayer::scale_native_rect

livecodeian committed:
[[ WindowsPlayer ]] Allow native view to occupy part to control rect

livecodeian committed:
[[ WindowsPlayer ]] Move MCPlatformPlayer class definition to platform-internal.h
[[ WindowsPlayer ]] Rename mac-player.mm to platform-player.cpp and remove mac-specific includes.

livecodeian committed:
[[ WindowsPlayer ]] Move handling of player-legacy vs player-platform to build configuration files.

livecodeian committed:
[[ WindowsPlayer ]] Add new sources to build config files and reorganise exclusions to allow Win32 builds to include required MCPlatform sources

livecodeian committed:
[[ WindowsPlayer ]] Add ifdefs to only include required MCPlatform features, and enable platform player feature on win32

livecodeian committed:
[[ WindowsPlayer ]] Remove QT-specific code from win32/platform-agnostic sources

livecodeian committed:
[[ WindowsPlayer ]] Remove code for win32 MCI player implementation

livecodeian committed:
[[ WindowsPlayer ]] Add compatibility MCPlatformWaitForEvent function

livecodeian committed:
[[ Windows ]] Add Windows implementation of MCPlatformCreatePlayer, returning an MCWin32DSPlayer ref

livecodeian committed:
[[ WindowsPlayer ]] Fix wrong type given for MCPlatformPlayer Url & FileName properties

livecodeian committed:
[[ WindowsPlayer ]] Clean up resources when closing.
[[ WIndowsPlayer ]]  Add video window proc to reconfigure video window on resize.
[[ WindowsPlayer ]] Fix video window class not registered

livecodeian committed:
[[ WIndowsPlayer ]] Replace MCPlayer::changewindow() call with detach+reattach native view.
[[ WindowsPlayer ]] Fix infinite loop when MCPlayer::playstop() returns false due to being uninitialised

livecodeian committed:
[[ WindowsPlayer ]] Fix video window being destroyed with native layer wrapper

livecodeian committed:
[[ WindowsPlayer ]] Don't attach / detach video window when containing view shown / hidden
[[ WindowsPlayer ]] Add property accessor stubs

livecodeian committed:
[[ WindowsPlayer ]] Fix result checking for functions that can return S_FALSE

livecodeian committed:
[[ WindowsPlayer ]] Implement startTime & finishTime properties

livecodeian committed:
[[ WindowsPlayer ]] Implement looping property

livecodeian committed:
[[ WindowsPlayer ]] Enable MCStringConvertToBSTR function and add library dependency to libfoundation.gyp

livecodeian committed:
[[ WindowsPlayer ]] Use MCAutoCustomPointer instead of adding YAARC (Yet Another AutoRelease Class)

livecodeian committed:
[[ WindowsPlayer ]] Implement currentTimeChanged callback message

livecodeian committed:
[[ WindowsPlayer ]] Convert between percentage and dB level when getting / setting volume

livecodeian committed:
[[ WindowsPlayer ]] Implement playSelection property

livecodeian committed:
[[ WindowsPlayer ]] Fix popup not dismissing and blocking events

livecodeian committed:
[[ WindowsPlayer ]] Implement loadedTime property

livecodeian committed:
[[ WindowsPlayer ]] Add deprecation note to QTVR property docs

livecodeian committed:
[[ WindowsPlayer ]] Insert copyright notice

livecodeian committed:
[[ WindowsPlayer ]] Update player property docs to indicate features not currently supported on Windows

livecodeian committed:
Merge branch 'develop' of https://github.com/livecode/livecode into feature-windows_multimedia

livecodeian committed:
Merge branch 'feature-windows_multimedia' of https://github.com/livecode/livecode-features into feature-windows_multimedia

peter-b commented:
Do you think it would be a good idea to add the version in which each dictionary entry became deprecated?  I often find that that's a useful piece of information to have (but that might just be me).

I also note that you've marked certain dictionary entries as no longer available on Windows without either marking them as deprecated or adding a "Changes" section.

It would be good to double-check that files you've significantly modified have their copyright headers updated to "2016".

<details><summary>

<i>Previously, livecodeian (Ian Macphail) wrote…</i></summary>



> #### [[ WindowsPlayer ]] Update the player object on Windows to use DirectShow
> 
> Modified MCPlatformPlayer to return the native view, using MCNativeLayer to manage geometry / visibility changes.
> 
> Implemented new MCPlatformPlayer based on DirectShow - switched Windows engine to use MCPlayer based on MCPlatformPlayer.
> 
> </details>

---

Reviewed 13 of 22 files at r1, 12 of 22 files at r2, 27 of 27 files at r3.
Review status: 52 of 55 files reviewed at latest revision, 20 unresolved discussions.

---

_[docs/dictionary/property/callbacks.lcdoc, line 13 [r3]](https://reviewable.io:443/reviews/livecode/livecode/4022#-KHy7cM8dMozuWDUIyB_:-KHy7cM8dMozuWDUIyBa:2128615132) ([raw file](https://github.com/livecode/livecode/blob/ebedce7c99dd9e84f87dacd0399faaa39c3f21d8/docs/dictionary/property/callbacks.lcdoc#L13)):_

> ```
> Introduced: 1.0
> 
> OS: mac,linux
> ```

Do I need to add a backlog item to implement this for Windows? Or is it infeasible to implement this for the new Windows player?

---

_[docs/dictionary/property/enabledTracks.lcdoc, line 13 [r3]](https://reviewable.io:443/reviews/livecode/livecode/4022#-KHy8DWfbDHsz6m_mCKo:-KHy8DWgp4ZS7Kf13zrO:468701807) ([raw file](https://github.com/livecode/livecode/blob/ebedce7c99dd9e84f87dacd0399faaa39c3f21d8/docs/dictionary/property/enabledTracks.lcdoc#L13)):_

> ```
> Introduced: 1.0
> 
> OS: mac,linux
> ```

FYI I already have a backlog item for implementing track listing/selection on Windows.

---

_[docs/dictionary/property/mediaTypes.lcdoc, line 13 [r3]](https://reviewable.io:443/reviews/livecode/livecode/4022#-KHy8M0dFfv_POcuEbP2:-KHy8M0dFfv_POcuEbP3:-1230672) ([raw file](https://github.com/livecode/livecode/blob/ebedce7c99dd9e84f87dacd0399faaa39c3f21d8/docs/dictionary/property/mediaTypes.lcdoc#L13)):_

> ```
> Introduced: 1.0
> 
> OS: mac,linux,ios,android
> ```

FYI I already have a backlog item for implementing `mediaTypes` for the Windows player

---

_[engine/src/mac-qt-player.mm, line 284 [r3]](https://reviewable.io:443/reviews/livecode/livecode/4022#-KHyE9Qv3jVW2aAwlVAw:-KHyE9Qv3jVW2aAwlVAx:-327000806) ([raw file](https://github.com/livecode/livecode/blob/ebedce7c99dd9e84f87dacd0399faaa39c3f21d8/engine/src/mac-qt-player.mm#L284)):_

> ``` XML
>   r_view = m_view;
>   return true;
> }
> ```

If I interpret the code correctly, each of the player implementation classes includes this same method almost verbatim.  Is there a missing superclass or mixin?

---

_[engine/src/mac-qt-player.mm, line 966 [r3]](https://reviewable.io:443/reviews/livecode/livecode/4022#-KHyA90YOjbHhlHfmoxN:-KHyA90YOjbHhlHfmoxO:-913127626) ([raw file](https://github.com/livecode/livecode/blob/ebedce7c99dd9e84f87dacd0399faaa39c3f21d8/engine/src/mac-qt-player.mm#L966)):_

> ``` XML
> ////////////////////////////////////////////////////////
> 
> extern bool MCQTInitialize();
> ```

Does it make sense for this to be declared in a header file?

---

_[engine/src/platform-internal.h, line 240 [r3]](https://reviewable.io:443/reviews/livecode/livecode/4022#-KHyBfqn46Tg2VFE9yTo:-KHyBfqn46Tg2VFE9yTp:-1663937917) ([raw file](https://github.com/livecode/livecode/blob/ebedce7c99dd9e84f87dacd0399faaa39c3f21d8/engine/src/platform-internal.h#L240)):_

> ``` objc
> 
>   void Retain(void);
>   void Release(void);
> ```

BTW at some point in would be good to get MCPlatformPlayer using the refcounted mixin that @livecodefraser implemented during the LiveCode 8 clipboard refactor.

---

_[engine/src/platform.h, line 1026 [r3]](https://reviewable.io:443/reviews/livecode/livecode/4022#-KHyCgeixHkPTRUCpn8C:-KHyCgeixHkPTRUCpn8D:-484433450) ([raw file](https://github.com/livecode/livecode/blob/ebedce7c99dd9e84f87dacd0399faaa39c3f21d8/engine/src/platform.h#L1026)):_

> ``` objc
> void MCPlatformStopPlayer(MCPlatformPlayerRef player);
> 
> bool MCPlatformLockPlayerBitmap(MCPlatformPlayerRef player, uint32_t p_width, uint32_t p_height, MCImageBitmap*& r_bitmap);
> ```

BTW is there a suitable "dimension" struct type that could be used here?

---

_[engine/src/player-platform.cpp, line 3476 [r3]](https://reviewable.io:443/reviews/livecode/livecode/4022#-KHyD6jQATk3TWLhHZck:-KHyD6jQATk3TWLhHZcl:-17744666) ([raw file](https://github.com/livecode/livecode/blob/ebedce7c99dd9e84f87dacd0399faaa39c3f21d8/engine/src/player-platform.cpp#L3476)):_

> ``` cpp
> 
>   if (getflag(F_SHOW_BORDER))
>       t_rect = MCU_reduce_rect(t_rect, borderwidth);
> ```

BTW am I right in thinking that the border is shown around the _whole_ player including the controls?

---

_[engine/src/sysdefs.h, line 53 [r3]](https://reviewable.io:443/reviews/livecode/livecode/4022#-KHyF-VBlVxpX4gP1dxm:-KHyF-VCCXrKeTFp0vPd:1492250682) ([raw file](https://github.com/livecode/livecode/blob/ebedce7c99dd9e84f87dacd0399faaa39c3f21d8/engine/src/sysdefs.h#L53)):_

> ``` objc
> #define FEATURE_PLATFORM_APPLICATION
> #define FEATURE_PLATFORM_PLAYER
> #define FEATURE_PLATFORM_WINDOW
> ```

Could you please add some comments that explain what `FEATURE_PLATFORM_APPLICATION` and `FEATURE_PLATFORM_WINDOW` imply? Unfortunately it's not obvious from inspection.

---

_[engine/src/w32-ds-player.cpp, line 1 [r3]](https://reviewable.io:443/reviews/livecode/livecode/4022#-KHyFEuyOHwhFANfavUu:-KHyFEuyOHwhFANfavUv:1151157784) ([raw file](https://github.com/livecode/livecode/blob/ebedce7c99dd9e84f87dacd0399faaa39c3f21d8/engine/src/w32-ds-player.cpp#L1)):_

> ``` cpp
> /* TODO - insert copyright notice */
> ```

Please insert the copyright notice.

---

_[engine/src/w32-ds-player.cpp, line 89 [r3]](https://reviewable.io:443/reviews/livecode/livecode/4022#-KHyGRrTlBbtA3lpylKv:-KHyGRrTlBbtA3lpylKw:-1244518256) ([raw file](https://github.com/livecode/livecode/blob/ebedce7c99dd9e84f87dacd0399faaa39c3f21d8/engine/src/w32-ds-player.cpp#L89)):_

> ``` cpp
>   bool GetFinishPosition(uint32_t &r_position);
>   bool SetFinishPosition(uint32_t p_position);
>   bool GetLoadedPosition(uint32_t &r_position);
> ```

The position parameters have units (frames? seconds?) but it's not immediately obvious what they are.  It might help to introduce a unit-based `typedef` and some comments explaining what's going on.

---

_[engine/src/w32-ds-player.cpp, line 136 [r3]](https://reviewable.io:443/reviews/livecode/livecode/4022#-KHyIy7HsoG198L59LlN:-KHyIy7HsoG198L59LlO:-303873421) ([raw file](https://github.com/livecode/livecode/blob/ebedce7c99dd9e84f87dacd0399faaa39c3f21d8/engine/src/w32-ds-player.cpp#L136)):_

> ``` cpp
> 
> #define WM_GRAPHNOTIFY (WM_APP + 1)   // Window message for graph events
> long FAR PASCAL DSHiddenWindowProc (HWND hWnd, UINT message, UINT wParam, LONG lParam)
> ```

This should be:

``` c++
LRESULT CALLBACK DSHiddenWindowProc(HWND hWnd, UINT message, WPARAM wParam, LPARAM lParam)
```

---

_[engine/src/w32-ds-player.cpp, line 154 [r3]](https://reviewable.io:443/reviews/livecode/livecode/4022#-KHyGhtt2TFO3umrN4I8:-KHyGhtuXOY_sGJa5sKV:1864546855) ([raw file](https://github.com/livecode/livecode/blob/ebedce7c99dd9e84f87dacd0399faaa39c3f21d8/engine/src/w32-ds-player.cpp#L154)):_

> ``` cpp
>   //case WM_MOUSEMOVE:
>   //  DispatchMetaCardMessage("XMediaPlayer_MouseMove","");
>   //  break;
> ```

Please remove this code instead of commenting it out.

---

_[engine/src/w32-ds-player.cpp, line 185 [r3]](https://reviewable.io:443/reviews/livecode/livecode/4022#-KHyJ5fSCVn__TP4B8tx:-KHyJ5fT9RrpHXlzyAhb:1948614072) ([raw file](https://github.com/livecode/livecode/blob/ebedce7c99dd9e84f87dacd0399faaa39c3f21d8/engine/src/w32-ds-player.cpp#L185)):_

> ``` cpp
>       // Register the Monitor child window class
>       wc.style         = CS_HREDRAW|CS_VREDRAW|CS_OWNDC|CS_SAVEBITS;
>       wc.lpfnWndProc   = (WNDPROC)DSHiddenWindowProc;
> ```

You should be able to remove the cast to `WNDPROC` from this line once the definition of `DSHiddenWindowProc` is fixed.

---

_[engine/src/w32-ds-player.cpp, line 200 [r3]](https://reviewable.io:443/reviews/livecode/livecode/4022#-KHyHgLgulgcvWuwU78d:-KHyHgLgulgcvWuwU78e:494356547) ([raw file](https://github.com/livecode/livecode/blob/ebedce7c99dd9e84f87dacd0399faaa39c3f21d8/engine/src/w32-ds-player.cpp#L200)):_

> ``` cpp
> }
> 
> long FAR PASCAL DSVideoWindowProc(HWND hWnd, UINT message, UINT wParam, LONG lParam)
> ```

The type of `lParam` should be `LPARAM`, not `LONG`.  `LPARAM` is pointer-sized, while `LONG` is always 32 bits.

I suspect that the full declaration should be:

``` c++
LRESULT CALLBACK DsVideoWindowProc(HWND hWnd, UINT message, WPARAM wParam, LPARAM lParam)
```

---

_[engine/src/w32-ds-player.cpp, line 426 [r3]](https://reviewable.io:443/reviews/livecode/livecode/4022#-KHyJuPCgYKNkFHoSr-D:-KHyJuPCgYKNkFHoSr-E:1168708289) ([raw file](https://github.com/livecode/livecode/blob/ebedce7c99dd9e84f87dacd0399faaa39c3f21d8/engine/src/w32-ds-player.cpp#L426)):_

> ``` cpp
>       return false;
> 
>   if (S_OK != t_video->put_Owner((OAHWND)p_hwnd))
> ```

Is `HWND` a supertype of `OAHWND`? If `HWND` to `OAHWND` is a downcast it might be worthwhile making it explicit with a `static_cast<OAHWND>`.

---

_[engine/src/w32-ds-player.cpp, line 605 [r3]](https://reviewable.io:443/reviews/livecode/livecode/4022#-KHyKu4bpBgWRPljDYjs:-KHyKu4bpBgWRPljDYjt:767880468) ([raw file](https://github.com/livecode/livecode/blob/ebedce7c99dd9e84f87dacd0399faaa39c3f21d8/engine/src/w32-ds-player.cpp#L605)):_

> ``` cpp
>       return false;
> 
>   r_time_scale = (uint32_t)(10000 / t_media_time); // convert to units/second
> ```

Would it be better to use an explicit rounding function here?

---

_[engine/src/w32-ds-player.cpp, line 618 [r3]](https://reviewable.io:443/reviews/livecode/livecode/4022#-KHyL1WD7cKDHCWzfJWJ:-KHyL1WD7cKDHCWzfJWK:-191996406) ([raw file](https://github.com/livecode/livecode/blob/ebedce7c99dd9e84f87dacd0399faaa39c3f21d8/engine/src/w32-ds-player.cpp#L618)):_

> ``` cpp
>       return false;
> 
>   r_duration = (uint32_t)t_duration;
> ```

Missing overflow check (64-bit signed to 32-bit unsigned) -- etc. for several other `LONGLONG` to `uint32_t` casts in this file.

---

_[engine/src/w32-ds-player.cpp, line 1081 [r3]](https://reviewable.io:443/reviews/livecode/livecode/4022#-KHyLqpZxREe0z0Ym-P4:-KHyLqpZxREe0z0Ym-P5:-1961341495) ([raw file](https://github.com/livecode/livecode/blob/ebedce7c99dd9e84f87dacd0399faaa39c3f21d8/engine/src/w32-ds-player.cpp#L1081)):_

> ``` cpp
>   case kMCPlatformPlayerPropertyMediaTypes:
>   case kMCPlatformPlayerPropertyURL:
>       MCLog("UNIMPLEMENTED");
> ```

Would it be appropriate to set `r_value` to an appropriate "default" value and/or raise a script execution error here?

---

_[engine/src/w32stack.cpp, line 339 [r3]](https://reviewable.io:443/reviews/livecode/livecode/4022#-KHyMYNdAyZZLU4XGeOF:-KHyMYNee16yCbYA12af:341235575) ([raw file](https://github.com/livecode/livecode/blob/ebedce7c99dd9e84f87dacd0399faaa39c3f21d8/engine/src/w32stack.cpp#L339)):_

> ``` cpp
>                       t_layer->OnAttach();
>                   }
>               }
> ```

Does it make sense to loop over "all controls that have native layers" here rather than "all players"? Are browsers being handled properly here for example?

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/livecode/livecode/4022#-:-KHy8YQHJ1O2oqPCZrRJ:-1127292824)_

<!-- Sent from Reviewable.io -->

livecodeian commented:
_[engine/src/mac-qt-player.mm, line 284 [r3]](https://reviewable.io:443/reviews/livecode/livecode/4022#-KHyE9Qv3jVW2aAwlVAw:-KHyU9W2V4iH-suAZQG4:59672367) ([raw file](https://github.com/livecode/livecode/blob/ebedce7c99dd9e84f87dacd0399faaa39c3f21d8/engine/src/mac-qt-player.mm#L284)):_
<details><summary>

<i>Previously, peter-b (Peter TB Brett) wrote…</i></summary>



> If I interpret the code correctly, each of the player implementation classes includes this same method almost verbatim.  Is there a missing superclass or mixin?</details>
> Each class defines m_view differently. The only reason the code is the same here is that the different types can all be cast to void.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/livecode/livecode/4022)_

<!-- Sent from Reviewable.io -->

livecodeian committed:
[[ WindowsPlayer ]] Fix window proc signatures and remove unused code

livecodeian commented:
Review status: 52 of 55 files reviewed at latest revision, 15 unresolved discussions.

---

_[docs/dictionary/property/callbacks.lcdoc, line 13 [r3]](https://reviewable.io:443/reviews/livecode/livecode/4022#-KHy7cM8dMozuWDUIyB_:-KHyVpuFNzyRG8JzValf:135278855) ([raw file](https://github.com/livecode/livecode/blob/ebedce7c99dd9e84f87dacd0399faaa39c3f21d8/docs/dictionary/property/callbacks.lcdoc#L13)):_
<details><summary>

<i>Previously, peter-b (Peter TB Brett) wrote…</i></summary>



> Do I need to add a backlog item to implement this for Windows? Or is it infeasible to implement this for the new Windows player?</details>
> It should be possible to add in - there's already a timer event to track currenttime while playing. That can be expanded to check currenttime against the list of callbacks.

---

_[engine/src/player-platform.cpp, line 3476 [r3]](https://reviewable.io:443/reviews/livecode/livecode/4022#-KHyD6jQATk3TWLhHZck:-KHyXHRzA4L5TLRS-elP:147300163) ([raw file](https://github.com/livecode/livecode/blob/ebedce7c99dd9e84f87dacd0399faaa39c3f21d8/engine/src/player-platform.cpp#L3476)):_
<details><summary>

<i>Previously, peter-b (Peter TB Brett) wrote…</i></summary>



> BTW am I right in thinking that the border is shown around the _whole_ player including the controls?</details>
> Yes, I made sure to preserve the current behaviour.

---

_[engine/src/w32-ds-player.cpp, line 1 [r3]](https://reviewable.io:443/reviews/livecode/livecode/4022#-KHyFEuyOHwhFANfavUu:-KHyXaFWXBxDpS8vwnNH:1120488839) ([raw file](https://github.com/livecode/livecode/blob/ebedce7c99dd9e84f87dacd0399faaa39c3f21d8/engine/src/w32-ds-player.cpp#L1)):_
<details><summary>

<i>Previously, peter-b (Peter TB Brett) wrote…</i></summary>



> Please insert the copyright notice.</details>
> I had pushed out a branch that was behind by 1 commit - have fixed this now.

---

_[engine/src/w32-ds-player.cpp, line 426 [r3]](https://reviewable.io:443/reviews/livecode/livecode/4022#-KHyJuPCgYKNkFHoSr-D:-KHynYt0I4hxRj96EAOa:229410440) ([raw file](https://github.com/livecode/livecode/blob/ebedce7c99dd9e84f87dacd0399faaa39c3f21d8/engine/src/w32-ds-player.cpp#L426)):_
<details><summary>

<i>Previously, peter-b (Peter TB Brett) wrote…</i></summary>



> Is `HWND` a supertype of `OAHWND`? If `HWND` to `OAHWND` is a downcast it might be worthwhile making it explicit with a `static_cast<OAHWND>`.</details>
> OAHWND is just a typedef for LONG_PTR. This appears to be standard Windows boilerplate for passing a HWND to an OLE object.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/livecode/livecode/4022)_

<!-- Sent from Reviewable.io -->

livecodeian committed:
[[ WindowsPlayer ]] Reattach all controls with native layers instead of just players when switching windows
[[ WindowsPlayer ]] Use previously set rect & viewport rect when attaching native layer

livecodeian commented:
Review status: 52 of 55 files reviewed at latest revision, 15 unresolved discussions.

---

_[engine/src/w32-ds-player.cpp, line 89 [r3]](https://reviewable.io:443/reviews/livecode/livecode/4022#-KHyGRrTlBbtA3lpylKv:-KHzFOW7ytQkOUaO5Dzc:657503268) ([raw file](https://github.com/livecode/livecode/blob/ebedce7c99dd9e84f87dacd0399faaa39c3f21d8/engine/src/w32-ds-player.cpp#L89)):_
<details><summary>

<i>Previously, peter-b (Peter TB Brett) wrote…</i></summary>



> The position parameters have units (frames? seconds?) but it's not immediately obvious what they are.  It might help to introduce a unit-based `typedef` and some comments explaining what's going on.</details>
> The units depend on the time format used by the filter graph, as reported by IMediaSeeking->GetTimeFormat() so can't be known ahead of time. That's the point of the GetTimeScale method, which returns the number of units per second. I'm not sure what more I can do here.

---

_[engine/src/w32-ds-player.cpp, line 618 [r3]](https://reviewable.io:443/reviews/livecode/livecode/4022#-KHyL1WD7cKDHCWzfJWJ:-KHzHOoFCqmujoU6YBTK:-1450344385) ([raw file](https://github.com/livecode/livecode/blob/ebedce7c99dd9e84f87dacd0399faaa39c3f21d8/engine/src/w32-ds-player.cpp#L618)):_
<details><summary>

<i>Previously, peter-b (Peter TB Brett) wrote…</i></summary>



> Missing overflow check (64-bit signed to 32-bit unsigned) -- etc. for several other `LONGLONG` to `uint32_t` casts in this file.</details>
> I'll add the checks in. Assuming reference time units (100 nanoseconds) the maximum duration that can be stored in a uint32_t is ~119.3 hours, or just shy of 5 days :) Let's hope none of our users have video files that long.

---

_[engine/src/w32stack.cpp, line 339 [r3]](https://reviewable.io:443/reviews/livecode/livecode/4022#-KHyMYNdAyZZLU4XGeOF:-KHzE8bIDce5ITvhTdXr:-110268197) ([raw file](https://github.com/livecode/livecode/blob/ebedce7c99dd9e84f87dacd0399faaa39c3f21d8/engine/src/w32stack.cpp#L339)):_
<details><summary>

<i>Previously, peter-b (Peter TB Brett) wrote…</i></summary>



> Does it make sense to loop over "all controls that have native layers" here rather than "all players"? Are browsers being handled properly here for example?</details>
> I've pushed a new commit that does as you suggest - this will affect all controls with native layers, not just players

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/livecode/livecode/4022)_

<!-- Sent from Reviewable.io -->

livecodeian commented:
Review status: 52 of 55 files reviewed at latest revision, 15 unresolved discussions.

---

_[engine/src/mac-qt-player.mm, line 966 [r3]](https://reviewable.io:443/reviews/livecode/livecode/4022#-KHyA90YOjbHhlHfmoxN:-KHzWFPwpB8QUTL9IBSg:97507111) ([raw file](https://github.com/livecode/livecode/blob/ebedce7c99dd9e84f87dacd0399faaa39c3f21d8/engine/src/mac-qt-player.mm#L966)):_
<details><summary>

<i>Previously, peter-b (Peter TB Brett) wrote…</i></summary>



> Does it make sense for this to be declared in a header file?</details>
> With QuickTime support deprecated I don't think it's worth spending the time to clean up the header declarations.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/livecode/livecode/4022)_

<!-- Sent from Reviewable.io -->

peter-b commented:
Reviewed 1 of 22 files at r2, 2 of 3 files at r4.
Review status: 54 of 55 files reviewed at latest revision, 7 unresolved discussions.

---

_[docs/dictionary/property/callbacks.lcdoc, line 13 [r3]](https://reviewable.io:443/reviews/livecode/livecode/4022#-KHy7cM8dMozuWDUIyB_:-KI1uFfXP7XyYW_ZlJPt:-11008869) ([raw file](https://github.com/livecode/livecode/blob/ebedce7c99dd9e84f87dacd0399faaa39c3f21d8/docs/dictionary/property/callbacks.lcdoc#L13)):_
<details><summary>

<i>Previously, livecodeian (Ian Macphail) wrote…</i></summary>



> It should be possible to add in - there's already a timer event to track currenttime while playing. That can be expanded to check currenttime against the list of callbacks.
> </details>
> Okay, cool -- I'll add it to the backlog.

---

_[engine/src/mac-qt-player.mm, line 284 [r3]](https://reviewable.io:443/reviews/livecode/livecode/4022#-KHyE9Qv3jVW2aAwlVAw:-KI1uZ-GJVEgyljfGg8J:744016330) ([raw file](https://github.com/livecode/livecode/blob/ebedce7c99dd9e84f87dacd0399faaa39c3f21d8/engine/src/mac-qt-player.mm#L284)):_
<details><summary>

<i>Previously, livecodeian (Ian Macphail) wrote…</i></summary>



> Each class defines m_view differently. The only reason the code is the same here is that the different types can all be cast to void.
> </details>
> OK. Is it worth considering a `NativeControl<ViewType>` template class?

---

_[engine/src/mac-qt-player.mm, line 966 [r3]](https://reviewable.io:443/reviews/livecode/livecode/4022#-KHyA90YOjbHhlHfmoxN:-KI1ud-NZJNpJAj-lcyT:1700654871) ([raw file](https://github.com/livecode/livecode/blob/ebedce7c99dd9e84f87dacd0399faaa39c3f21d8/engine/src/mac-qt-player.mm#L966)):_
<details><summary>

<i>Previously, livecodeian (Ian Macphail) wrote…</i></summary>



> With QuickTime support deprecated I don't think it's worth spending the time to clean up the header declarations.
> </details>
> OK, fair enough.

---

_[engine/src/w32-ds-player.cpp, line 1 [r3]](https://reviewable.io:443/reviews/livecode/livecode/4022#-KHyFEuyOHwhFANfavUu:-KI1uj1IyNADtt6i8rIE:1396757992) ([raw file](https://github.com/livecode/livecode/blob/ebedce7c99dd9e84f87dacd0399faaa39c3f21d8/engine/src/w32-ds-player.cpp#L1)):_
<details><summary>

<i>Previously, livecodeian (Ian Macphail) wrote…</i></summary>



> I had pushed out a branch that was behind by 1 commit - have fixed this now.
> </details>
> The usual practice in GPL headers when there is shared copyright is to write something like:

```
Copyright (C) 2016 LiveCode Ltd.
Copyright (C) 2016 Docas AG.
```

I think that this is fine how it is however, thank you!

---

_[engine/src/w32-ds-player.cpp, line 89 [r3]](https://reviewable.io:443/reviews/livecode/livecode/4022#-KHyGRrTlBbtA3lpylKv:-KI1v3tyjNvZP4tHIKHx:1083774781) ([raw file](https://github.com/livecode/livecode/blob/ebedce7c99dd9e84f87dacd0399faaa39c3f21d8/engine/src/w32-ds-player.cpp#L89)):_
<details><summary>

<i>Previously, livecodeian (Ian Macphail) wrote…</i></summary>



> The units depend on the time format used by the filter graph, as reported by IMediaSeeking->GetTimeFormat() so can't be known ahead of time. That's the point of the GetTimeScale method, which returns the number of units per second. I'm not sure what more I can do here.
> </details>
> In the graphics library there's an `MCGFloat` type which abstracts whether the graphics are being calculated using single- or double-precision floating point arithmetic -- and it's used to construct all the other graphics geometry types.

It's probably not necessary to do anything about this; I was just curious whether adding `typedef uint32_t MCMediaPosition` would make the code clearer, but from what you say perhaps it wouldn't make any difference.

---

_[engine/src/w32-ds-player.cpp, line 426 [r3]](https://reviewable.io:443/reviews/livecode/livecode/4022#-KHyJuPCgYKNkFHoSr-D:-KI1vpi1Zi7sDZg9Zvi-:-2044855707) ([raw file](https://github.com/livecode/livecode/blob/ebedce7c99dd9e84f87dacd0399faaa39c3f21d8/engine/src/w32-ds-player.cpp#L426)):_
<details><summary>

<i>Previously, livecodeian (Ian Macphail) wrote…</i></summary>



> OAHWND is just a typedef for LONG_PTR. This appears to be standard Windows boilerplate for passing a HWND to an OLE object.
> </details>
> Thanks, that makes sense!

---

_[engine/src/w32-ds-player.cpp, line 618 [r3]](https://reviewable.io:443/reviews/livecode/livecode/4022#-KHyL1WD7cKDHCWzfJWJ:-KI1vwvUoeNBVjgVqDYU:1841733026) ([raw file](https://github.com/livecode/livecode/blob/ebedce7c99dd9e84f87dacd0399faaa39c3f21d8/engine/src/w32-ds-player.cpp#L618)):_
<details><summary>

<i>Previously, livecodeian (Ian Macphail) wrote…</i></summary>



> I'll add the checks in. Assuming reference time units (100 nanoseconds) the maximum duration that can be stored in a uint32_t is ~119.3 hours, or just shy of 5 days :) Let's hope none of our users have video files that long.
> </details>
> Knowing our users, I wouldn't put it past them! I do think that the overflow checks are necessary. Do you think it makes sense to make the conversion saturating (i.e. if the duration is longer than `UINT32_MAX`, convert it to `UINT32_MAX`? Or is it better to just fail and return `false`?

---

_[engine/src/w32stack.cpp, line 339 [r3]](https://reviewable.io:443/reviews/livecode/livecode/4022#-KHyMYNdAyZZLU4XGeOF:-KI1wYZJNB1GSK3_JlWJ:-1951500724) ([raw file](https://github.com/livecode/livecode/blob/ebedce7c99dd9e84f87dacd0399faaa39c3f21d8/engine/src/w32stack.cpp#L339)):_
<details><summary>

<i>Previously, livecodeian (Ian Macphail) wrote…</i></summary>



> I've pushed a new commit that does as you suggest - this will affect all controls with native layers, not just players
> </details>
> LGTM, thank  you.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/livecode/livecode/4022)_

<!-- Sent from Reviewable.io -->

livecodeian committed:
[[ WindowsPlayer ]] Use MCGIntegerSize for player snapshot dimensions

peter-b commented:
Reviewed 6 of 7 files at r5.
Review status: 54 of 55 files reviewed at latest revision, 4 unresolved discussions.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/livecode/livecode/4022)_

<!-- Sent from Reviewable.io -->

livecodeian commented:
Review status: 54 of 55 files reviewed at latest revision, 4 unresolved discussions.

---

_[docs/dictionary/property/callbacks.lcdoc, line 13 [r3]](https://reviewable.io:443/reviews/livecode/livecode/4022#-KHy7cM8dMozuWDUIyB_:-KI25h6xa11JRyBkqpqh:1249236387) ([raw file](https://github.com/livecode/livecode/blob/ebedce7c99dd9e84f87dacd0399faaa39c3f21d8/docs/dictionary/property/callbacks.lcdoc#L13)):_
<details><summary>

<i>Previously, peter-b (Peter TB Brett) wrote…</i></summary>



> Okay, cool -- I'll add it to the backlog.
> </details>
> Just a note that there may be an IReferenceClock within the filter graph that can be used instead

---

_[engine/src/mac-qt-player.mm, line 284 [r3]](https://reviewable.io:443/reviews/livecode/livecode/4022#-KHyE9Qv3jVW2aAwlVAw:-KI2sgK_QJCsLYUynjQz:1441589) ([raw file](https://github.com/livecode/livecode/blob/ebedce7c99dd9e84f87dacd0399faaa39c3f21d8/engine/src/mac-qt-player.mm#L284)):_
<details><summary>

<i>Previously, peter-b (Peter TB Brett) wrote…</i></summary>



> OK. Is it worth considering a `NativeControl<ViewType>` template class?
> </details>
> At the moment that seems unnecessary. It may be something to consider during further development of MCPlatformPlayer.

---

_[engine/src/platform.h, line 1026 [r3]](https://reviewable.io:443/reviews/livecode/livecode/4022#-KHyCgeixHkPTRUCpn8C:-KI2ehLOUjVBUvu-7dtq:-2047981912) ([raw file](https://github.com/livecode/livecode/blob/ebedce7c99dd9e84f87dacd0399faaa39c3f21d8/engine/src/platform.h#L1026)):_
<details><summary>

<i>Previously, peter-b (Peter TB Brett) wrote…</i></summary>



> BTW is there a suitable "dimension" struct type that could be used here?
> </details>
> Yup, I've swapped this out with MCGIntegerSize

---

_[engine/src/w32-ds-player.cpp, line 618 [r3]](https://reviewable.io:443/reviews/livecode/livecode/4022#-KHyL1WD7cKDHCWzfJWJ:-KI22ad2_mHL6XJkM09q:281809852) ([raw file](https://github.com/livecode/livecode/blob/ebedce7c99dd9e84f87dacd0399faaa39c3f21d8/engine/src/w32-ds-player.cpp#L618)):_
<details><summary>

<i>Previously, peter-b (Peter TB Brett) wrote…</i></summary>



> Knowing our users, I wouldn't put it past them! I do think that the overflow checks are necessary. Do you think it makes sense to make the conversion saturating (i.e. if the duration is longer than `UINT32_MAX`, convert it to `UINT32_MAX`? Or is it better to just fail and return `false`?
> </details>
> I'd rather it fail than return an incorrect value.

---

_[engine/src/w32-ds-player.cpp, line 1081 [r3]](https://reviewable.io:443/reviews/livecode/livecode/4022#-KHyLqpZxREe0z0Ym-P4:-KI239OpmjvpF0MlLvVr:1354945248) ([raw file](https://github.com/livecode/livecode/blob/ebedce7c99dd9e84f87dacd0399faaa39c3f21d8/engine/src/w32-ds-player.cpp#L1081)):_
<details><summary>

<i>Previously, peter-b (Peter TB Brett) wrote…</i></summary>



> Would it be appropriate to set `r_value` to an appropriate "default" value and/or raise a script execution error here?
> </details>
> I think reporting failure to the calling function is the way to go for unsupported properties, and raising exceptions when get/set of individual properties fails for whatever reason. The problem with that is that the calling functions were not written in a way to handle failures so revising this whole area will be a lot of work.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/livecode/livecode/4022)_

<!-- Sent from Reviewable.io -->

livecodeian committed:
[[ WindowsPlayer ]] Return false if fetched position property values will overflow uint32_t

livecodeian commented:
Review status: 54 of 55 files reviewed at latest revision, 5 unresolved discussions.

---

_[engine/src/w32-ds-player.cpp, line 605 [r3]](https://reviewable.io:443/reviews/livecode/livecode/4022#-KHyKu4bpBgWRPljDYjs:-KI2vcLdrl1ip65SXvg_:-513690470) ([raw file](https://github.com/livecode/livecode/blob/ebedce7c99dd9e84f87dacd0399faaa39c3f21d8/engine/src/w32-ds-player.cpp#L605)):_
<details><summary>

<i>Previously, peter-b (Peter TB Brett) wrote…</i></summary>



> Would it be better to use an explicit rounding function here?
> </details>
> For whatever reason the timeScale player property was defined as a positive integer, so the assumption here is that there will always be a whole number of interval units per second so no rounding required

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/livecode/livecode/4022)_

<!-- Sent from Reviewable.io -->

livecodeian commented:
Review status: 54 of 55 files reviewed at latest revision, 5 unresolved discussions.

---

_[engine/src/sysdefs.h, line 53 [r3]](https://reviewable.io:443/reviews/livecode/livecode/4022#-KHyF-VBlVxpX4gP1dxm:-KI2yb1EVxLNe2ymPIiQ:-1357307269) ([raw file](https://github.com/livecode/livecode/blob/ebedce7c99dd9e84f87dacd0399faaa39c3f21d8/engine/src/sysdefs.h#L53)):_
<details><summary>

<i>Previously, peter-b (Peter TB Brett) wrote…</i></summary>



> Could you please add some comments that explain what `FEATURE_PLATFORM_APPLICATION` and `FEATURE_PLATFORM_WINDOW` imply? Unfortunately it's not obvious from inspection.
> </details>
> Well, I added that so I could use the code in desktop.cpp and platform.cpp that handles the binding between MCPlatformPlayer and the MCPlayer control without having to go the whole hog and completely port the Windows engine to use MCPlatform. It might not be the best way to have done it. I'm wondering now if I should have moved the relevant code to platform-player.cpp and player-platform.cpp instead.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/livecode/livecode/4022)_

<!-- Sent from Reviewable.io -->

peter-b commented:
Review status: 54 of 55 files reviewed at latest revision, 4 unresolved discussions.

---

_[engine/src/platform.h, line 1026 [r3]](https://reviewable.io:443/reviews/livecode/livecode/4022#-KHyCgeixHkPTRUCpn8C:-KI3-AB8W_saAj8iTtt9:1683908099) ([raw file](https://github.com/livecode/livecode/blob/ebedce7c99dd9e84f87dacd0399faaa39c3f21d8/engine/src/platform.h#L1026)):_
<details><summary>

<i>Previously, livecodeian (Ian Macphail) wrote…</i></summary>



> Yup, I've swapped this out with MCGIntegerSize
> </details>
> OK, thanks!

---

_[engine/src/sysdefs.h, line 53 [r3]](https://reviewable.io:443/reviews/livecode/livecode/4022#-KHyF-VBlVxpX4gP1dxm:-KI3-Y7pt99X9HIMRddd:-985822282) ([raw file](https://github.com/livecode/livecode/blob/ebedce7c99dd9e84f87dacd0399faaa39c3f21d8/engine/src/sysdefs.h#L53)):_
<details><summary>

<i>Previously, livecodeian (Ian Macphail) wrote…</i></summary>



> Well, I added that so I could use the code in desktop.cpp and platform.cpp that handles the binding between MCPlatformPlayer and the MCPlayer control without having to go the whole hog and completely port the Windows engine to use MCPlatform. It might not be the best way to have done it. I'm wondering now if I should have moved the relevant code to platform-player.cpp and player-platform.cpp instead.
> </details>
> Would there be any other benefits to porting the Windows engine to MCPlatform?  Should I add it to the backlog?

---

_[engine/src/w32-ds-player.cpp, line 1081 [r3]](https://reviewable.io:443/reviews/livecode/livecode/4022#-KHyLqpZxREe0z0Ym-P4:-KI31wwEWFZQlouIScUw:1299454334) ([raw file](https://github.com/livecode/livecode/blob/ebedce7c99dd9e84f87dacd0399faaa39c3f21d8/engine/src/w32-ds-player.cpp#L1081)):_
<details><summary>

<i>Previously, livecodeian (Ian Macphail) wrote…</i></summary>



> I think reporting failure to the calling function is the way to go for unsupported properties, and raising exceptions when get/set of individual properties fails for whatever reason. The problem with that is that the calling functions were not written in a way to handle failures so revising this whole area will be a lot of work.
> </details>
> OK, I have added "sort out error propagation from player controls" to the backlog.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/livecode/livecode/4022)_

<!-- Sent from Reviewable.io -->

livecodeian committed:
[[ WindowsPlayer ]] Add changes section to docs for properties added to / removed from Windows player

livecodeian committed:
[[ WindowsPlayer ]] Add changes section to deprecated QTVR property docs

livecodeian committed:
[[ WindowsPlayer ]] Update copyright notice of significantly modified files.

livecodeian commented:
I have now added Changes: sections to the affected docs and updated Copyright notices where appropriate.

<details><summary>

<i>Previously, peter-b (Peter TB Brett) wrote…</i></summary>



> Do you think it would be a good idea to add the version in which each dictionary entry became deprecated?  I often find that that's a useful piece of information to have (but that might just be me).
> 
> I also note that you've marked certain dictionary entries as no longer available on Windows without either marking them as deprecated or adding a "Changes" section.
> 
> It would be good to double-check that files you've significantly modified have their copyright headers updated to "2016".
> </details>

---

Review status: 54 of 55 files reviewed at latest revision, 2 unresolved discussions.

---

_[engine/src/sysdefs.h, line 53 [r3]](https://reviewable.io:443/reviews/livecode/livecode/4022#-KHyF-VBlVxpX4gP1dxm:-KI3Es-bHKzl5sgMxkJS:1626223777) ([raw file](https://github.com/livecode/livecode/blob/ebedce7c99dd9e84f87dacd0399faaa39c3f21d8/engine/src/sysdefs.h#L53)):_
<details><summary>

<i>Previously, peter-b (Peter TB Brett) wrote…</i></summary>



> Would there be any other benefits to porting the Windows engine to MCPlatform?  Should I add it to the backlog?
> </details>
> You can add that to the backlog.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/livecode/livecode/4022#-:-KI3EiFZt0DXoFe7rDbG:-1692971960)_

<!-- Sent from Reviewable.io -->

livecodeian commented:
Review status: 54 of 55 files reviewed at latest revision, 2 unresolved discussions.

---

_[engine/src/w32-ds-player.cpp, line 618 [r3]](https://reviewable.io:443/reviews/livecode/livecode/4022#-KHyL1WD7cKDHCWzfJWJ:-KI3F5yudCNZ60SE7_wt:1737668118) ([raw file](https://github.com/livecode/livecode/blob/ebedce7c99dd9e84f87dacd0399faaa39c3f21d8/engine/src/w32-ds-player.cpp#L618)):_
<details><summary>

<i>Previously, livecodeian (Ian Macphail) wrote…</i></summary>



> I'd rather it fail than return an incorrect value.
> </details>
> I've added range checking to these now so overflow can be caught.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/livecode/livecode/4022)_

<!-- Sent from Reviewable.io -->

peter-b commented:
@livecodeian Thanks for addressing all those points! I think that all that's needed now is a nice comprehensive release note and this is ready to go!

<details><summary>

<i>Previously, livecodeian (Ian Macphail) wrote…</i></summary>



> I have now added Changes: sections to the affected docs and updated Copyright notices where appropriate.
> </details>

---

Reviewed 1 of 1 files at r6, 20 of 20 files at r7.
Review status: all files reviewed at latest revision, all discussions resolved.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/livecode/livecode/4022#-:-KI73yrrJfEbLZWtuHzP:-910085692)_

<!-- Sent from Reviewable.io -->

livecodeian committed:
[[ WindowsPlayer ]] Add release note.

livecodeian commented:
Release note added.

<details><summary>

<i>Previously, peter-b (Peter TB Brett) wrote…</i></summary>



> @livecodeian Thanks for addressing all those points! I think that all that's needed now is a nice comprehensive release note and this is ready to go!
> </details>

---

Review status: 55 of 56 files reviewed at latest revision, all discussions resolved.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/livecode/livecode/4022#-:-KI7hcv11xPhgbklK_ja:750077758)_

<!-- Sent from Reviewable.io -->

peter-b commented:
@livecode-vulcan review ok 37f1c3a

livecode-vulcan commented:
💙 review by @peter-b ok 37f1c3afaaea00acbad5b88d469ff5aec7fe78e7

livecode-vulcan commented:
😞 test failure 37f1c3afaaea00acbad5b88d469ff5aec7fe78e7
- try-community-linux-x86: [pending](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/2509)
- try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/2519)
- try-community-linux-x86_64: [pending](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/2504)
- try-community-emscripten: [pending](https://vulcan.livecode.com/builders/try-community-emscripten/builds/2520)
- try-community-ios: [failure](https://vulcan.livecode.com/builders/try-community-ios/builds/2526)
- try-community-win-x86: [pending](https://vulcan.livecode.com/builders/try-community-win-x86/builds/2530)
- try-community-android-armv6: [failure](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/2542)

livecodeian committed:
[[ WindowsPlayer ]] Exclude desktop & platform sources from mobile builds

peter-b commented:
@livecodeian Oops, I wish you'd told me you'd pushed these additional changes! We could have got them merged in today! :disappointed: 

@livecode-vulcan review ok 7a33cff

livecode-vulcan commented:
💙 review by @peter-b ok 7a33cff6dfdb2ea114730a70009a7312e74d0078

livecode-vulcan commented:
😎 test success 7a33cff6dfdb2ea114730a70009a7312e74d0078
- try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/2512)
- try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/2522)
- try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/2507)
- try-community-emscripten: [success](https://vulcan.livecode.com/builders/try-community-emscripten/builds/2523)
- try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/2529)
- try-community-win-x86: [success](https://vulcan.livecode.com/builders/try-community-win-x86/builds/2533)
- try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/2545)
